### PR TITLE
added views, controller, edited route and model for Categories

### PR DESCRIPTION
**Pull request by avminon**
app.php - added Collective\Html\HtmlServiceProvider::class under providers and Form, Html under aliases
routes.php - added Route::resource('categories', 'CategoryController');
UserController.php  - added construct function, edited index function
CategoryController.php - added controller for categories
Category.php - edited the Model to contain assign(), used for saving/updating category entries
categories/ - added views for categories

### CODE REVIEW

File: app/Category.php [hunk 1]
```diff
@@ -20,4 +20,20 @@ public function words()
         return $this->hasMany(Word::class, 'category_id');
     }
 
+    public function assign($values)
+    {
+
+        if($values->input('category_id') != null) {
```
Comment (nguyenthanhtung88) [inline app/Category.php]: Insert 1 space after `if`
(https://github.com/php-fig/fig-standards/blob/master/accepted/PSR-2-coding-style-guide.md#51-if-elseif-else)

Note: Check more `if` statement you used

File: app/Http/Controllers/CategoryController.php [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+{
+    protected $user;
+
+    public function __construct()
+    {
+        $this->middleware('auth');
+        $this->user = auth()->user();
+    }
+
+	public function index()
```
Comment (nguyenthanhtung88) [inline app/Http/Controllers/CategoryController.php]: The indent is not good

File: app/Http/Controllers/CategoryController.php [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+	{
+	    return view('categories.home', [
+            'categories' => Category::all(),
+            'user' => $this->user,
+            'title' => 'Categories'
+        ]);
+    }
+
```
Comment (nguyenthanhtung88) [inline app/Http/Controllers/CategoryController.php]: Remove empty line

File: app/Http/Controllers/CategoryController.php [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+    public function create()
+    {
+        return view('categories.add', [
+            'user' => $this->user,
+            'title' => 'Add category'
+        ]);
+    }
+
```
Comment (nguyenthanhtung88) [inline app/Http/Controllers/CategoryController.php]: Remove empty line

File: app/Http/Controllers/CategoryController.php [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+            return view('categories.edit', [
+                'category' => $category,
+                'title' => 'Edit category',
+                'user' => $this->user
+            ]);
+        } catch(ModelNotFoundException $e) {
+            \Session::flash('flash_error', 'Edit failed. The category cannot be found.');
+        }
+        return redirect('/categories');
+
```
Comment (nguyenthanhtung88) [inline app/Http/Controllers/CategoryController.php]: Remove this empty line

File: app/Http/Controllers/CategoryController.php [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+            'categories' => Category::all(),
+            'user' => $this->user,
+            'title' => 'Categories'
+        ]);
+    }
+
+
+    public function show($id)
+    {
+        return Category::find($id);
```
Comment (nguyenthanhtung88) [inline app/Http/Controllers/CategoryController.php]: Why we don't return view here?

File: app/Http/Controllers/UserController.php [hunk 7]
```diff
@@ -10,10 +10,20 @@
 
 class UserController extends Controller
 {
+    protected $user;
+
+    public function __construct()
+    {
+        $this->middleware('auth');
+        $this->user = \Auth::user();
```
Comment (nguyenthanhtung88) [inline app/Http/Controllers/UserController.php]: Use `auth()` helper here

File: app/Http/Controllers/CategoryController.php [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+use App\User;
+use Illuminate\Database\Eloquent\ModelNotFoundException;
+
+class CategoryController extends Controller
+{
+    protected $user;
+
+    public function __construct()
+    {
+        $this->middleware('auth');
```
Comment (nguyenthanhtung88) [inline app/Http/Controllers/CategoryController.php]: This `middleware` defined in `route` file, why do you define again here?

File: app/Http/Controllers/UserController.php [hunk 9]
```diff
@@ -10,10 +10,20 @@
 
 class UserController extends Controller
 {
+    protected $user;
+
+    public function __construct()
+    {
+        $this->middleware('auth');
```
Comment (nguyenthanhtung88) [inline app/Http/Controllers/UserController.php]: `Middleware` defined in `route` file

File: app/Http/Controllers/UserController.php [hunk 10]
```diff
@@ -10,10 +10,20 @@
 
 class UserController extends Controller
 {
+    protected $user;
```
Comment (nguyenthanhtung88) [inline app/Http/Controllers/UserController.php]: I saw you use `$user` twice, in `UserController` and `CategoryController`, why don't you define this variable in `Controller` base, so in all children controllers (like this controller), we can use from parent.

File: app/Http/Controllers/CategoryController.php [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+
+
+    public function show($id)
+    {
+        return Category::find($id);
+    }
+
+    public function create()
+    {
+        return view('categories.add', [
```
Comment (nguyenthanhtung88) [inline app/Http/Controllers/CategoryController.php]: Why don't you use `categories.create` view?

File: app/Http/routes.php [hunk 12]
```diff
@@ -11,16 +11,13 @@
...
 ]);
 
-
-Route::group(['middleware' => 'auth'], function () {
-
-    Route::get('/', ['as' => 'home', 'uses' => 'UserController@index']);
-    Route::get('/home', ['as' => 'home', 'uses' => 'UserController@index']);
-    
+Route::group(['middleware' => ['auth']], function() {
+    Route::get('/',['as' => 'home', 'uses' => 'UserController@index']);
```
Comment (nguyenthanhtung88) [inline app/Http/routes.php]: Insert one space after `'/',`

File: app/Http/routes.php [hunk 13]
```diff
@@ -11,16 +11,13 @@
...
+    Route::get('/home',['as' => 'home', 'uses' => 'UserController@index']);
```
Comment (nguyenthanhtung88) [inline app/Http/routes.php]: Insert one space after `'/',`

File: resources/views/categories/add.blade.php [hunk 14]
```diff
@@ -0,0 +1,30 @@
...
+@section('title')
+    @if (auth()->guest())
+        Welcome!
+    @endif
+    @if (auth()->check())
+        {{$title}}
+    @endif
+@endsection
+@section('content')
+	{!! Form::open(['method' => 'post', 'route' => 'categories.store', 'files' => 'true']) !!}
```
Comment (nguyenthanhtung88) [inline resources/views/categories/add.blade.php]: You're using `tab` here for indent, not `4 spaces`, please fix and check another views and this view too.

File: config/app.php [hunk 15]
```diff
@@ -136,6 +136,7 @@
         Illuminate\Translation\TranslationServiceProvider::class,
         Illuminate\Validation\ValidationServiceProvider::class,
         Illuminate\View\ViewServiceProvider::class,
+        Collective\Html\HtmlServiceProvider::class,
```
Comment (nguyenthanhtung88) [inline config/app.php]: You use this without adding package to `composer.json` ???
I saw on https://laravelcollective.com/docs/5.0/html, they said that:
```
Begin by installing this package through Composer. Edit your project's composer.json file to require laravelcollective/html.

"require": {
    "laravelcollective/html": "~5.0"
}
```

File: resources/views/categories/add.blade.php [hunk 16]
```diff
@@ -0,0 +1,30 @@
+@extends('app')
+@section('title')
+    @if (auth()->guest())
+        Welcome!
+    @endif
+    @if (auth()->check())
+        {{$title}}
+    @endif
```
Comment (nguyenthanhtung88) [inline resources/views/categories/add.blade.php]:
```
@if (auth()->guest())
        Welcome!
    @endif
    @if (auth()->check())
        {{$title}}
    @endif
```
1. This code repeat in other views? don't repeat your code by copy/paste => write down in partial
2. Your `CategoryController` defined that authenticated users can go through its actions (because of `Middleware` auth) => So why you have to check `guest` here +_+

File: config/app.php (hunk 15 above)
Comment (avminon) [inline config/app.php]: I updated my composer.json sir and updated my composer when I added the package

**Comment by nguyenthanhtung88** [comment]
@avminon : You've just used `auth middleware` for authenticating users right? How about `admin` role?
please remember that only `ADMIN` can create/update/delete category

File: app/Http/routes.php [hunk 17]
```diff
@@ -1,26 +1,12 @@
...
-
-
-Route::group(['middleware' => 'auth'], function () {
-
+Route::group(['middleware' => ['auth']], function() {
     Route::get('/', ['as' => 'home', 'uses' => 'UserController@index']);
     Route::get('/home', ['as' => 'home', 'uses' => 'UserController@index']);
-    
+    Route::resource('categories', 'CategoryController');
+    Route::resource('words', 'WordController');
```
Comment (avminon) [inline app/Http/routes.php]: oops sorry I forgot to omit this line. Though Route::resource('words', 'WordController'); will be added later for Words

**Comment by avminon** [comment]
image
image

File: app/Http/Controllers/CategoryController.php [hunk 18]
```diff
@@ -0,0 +1,81 @@
...
+class CategoryController extends Controller
+{
+    public function __construct()
+    {
+        $this->user = auth()->user();
+    }
+
+    public function index()
+    {
+        return view('categories.home', [
```
Comment (nguyenthanhtung88) [inline app/Http/Controllers/CategoryController.php]: `cateogires.index` view is better

File: app/Category.php [hunk 19]
```diff
@@ -20,4 +20,18 @@ public function words()
...
 
+    public function assign($values)
+    {
+        if (!is_null($values->input('category_id'))) {
+            $this->id = $values->input('category_id');
+        }
+        $this->name = $values->input('category_name');
+        $this->description = $values->input('category_desc');
+        if (!empty($values->file('category_image'))) {
+            $imageName = uniqid() . '.' . $values->file('category_image')->getClientOriginalExtension();
```
Comment (nguyenthanhtung88) [inline app/Category.php]: You don't validate the extension of file?
What happen if i upload other extension like `.csv` or `.xlsx` files?
I think you should research a little bit about validate file extensions, it's not take so much time :+1:

File: app/Http/Controllers/Controller.php [hunk 20]
```diff
@@ -10,4 +10,6 @@
 abstract class Controller extends BaseController
 {
     use AuthorizesRequests, DispatchesJobs, ValidatesRequests;
+
+    private $user;
```
Comment (nguyenthanhtung88) [inline app/Http/Controllers/Controller.php]: Oops do you think this `private` variable we can use in children class like `UserController`?

File: app/Http/Controllers/UserController.php [hunk 21]
```diff
@@ -5,15 +5,19 @@
...
 use App\User;
 use Redirect;
-
 use Illuminate\Http\Request;
 
 class UserController extends Controller
 {
+    public function __construct()
+    {
+        $this->user = auth()->user();
```
Comment (nguyenthanhtung88) [inline app/Http/Controllers/UserController.php]: You repeat this code, please write in parent `__construct()` function.
Note that each classes (ex: `UserController`, `CategoryController`) inherit from one class (ex: `Controller`), they always call `__construct()` function from parent class first => so you can use `$this->user` in every children classes.

File: config/app.php (hunk 15 above)
Comment (nguyenthanhtung88) [inline config/app.php]: Yes, it's OK now.

File: resources/views/categories/create.blade.php [hunk 22]
```diff
@@ -0,0 +1,24 @@
+@extends('app')
+@section('title')
+    {{$title}}
```
Comment (nguyenthanhtung88) [inline resources/views/categories/create.blade.php]: `{{ $title }}` => please use this

File: resources/views/categories/create.blade.php [hunk 23]
```diff
@@ -0,0 +1,24 @@
...
+@endsection
+@section('content')
+    {!! Form::open(['method' => 'post', 'route' => 'categories.store', 'files' => 'true']) !!}
+        {{ csrf_field() }}
+        <div class="form-group">
+            {!! Form::text('category_name', '', array(
```
Comment (nguyenthanhtung88) [inline resources/views/categories/create.blade.php]: use `[...]` instead of `array()`, it's new feature from PHP 5.4.
Please check other location you used.

File: resources/views/categories/edit.blade.php [hunk 24]
```diff
@@ -0,0 +1,49 @@
...
+                            'class' => 'form-control'))
+                        !!}
+                    </td>
+            </tr>
+            <tr>
+                <td>
+                    {!! Form::textarea('category_desc', $category->description, array(
+                        'required' => 'required',
+                        'class' => 'form-control'))
+                        !!}
```
Comment (nguyenthanhtung88) [inline resources/views/categories/edit.blade.php]: Wrong Indent

File: resources/views/categories/edit.blade.php [hunk 25]
```diff
@@ -0,0 +1,49 @@
...
+                </td>
+            </tr>
+                <td>
+                    {!! Form::file('category_image') !!}
+                </td>
+            <tr>
```
Comment (nguyenthanhtung88) [inline resources/views/categories/edit.blade.php]: Why open `<tr>` tag here ??? should move up, i think its location is in line 38.

File: resources/views/categories/edit.blade.php [hunk 26]
```diff
@@ -0,0 +1,49 @@
...
+            </tr>
+           </table>
```
Comment (nguyenthanhtung88) [inline resources/views/categories/edit.blade.php]: Wrong Indent

File: resources/views/categories/edit.blade.php [hunk 27]
```diff
@@ -0,0 +1,49 @@
...
+        {!! Form::submit('Save') !!}
+    {!! Form::close() !!}
```
Comment (nguyenthanhtung88) [inline resources/views/categories/edit.blade.php]: Wrong Indent

File: resources/views/categories/edit.blade.php [hunk 28]
```diff
@@ -0,0 +1,49 @@
...
+    {!! Form::open(['method' => 'get', 'route' => 'categories.index']) !!}
+        {!! Form::submit('Cancel') !!}
```
Comment (nguyenthanhtung88) [inline resources/views/categories/edit.blade.php]: Oops, why you have to create a new form inside this form.
I think your purpose is redirect to `categories.index` page when click "Cancel" :cry:
=> For simple we only need `link` tag here is enough

File: resources/views/categories/home.blade.php [hunk 29]
```diff
@@ -0,0 +1,62 @@
+@extends('app')
+@section('title')
+    {{$title}}
+@endsection
+@section('content')
+    @if ($user->isAdmin())
+        <p>
+            {!! Form::open(['method' => 'get', 'route' => 'categories.create']) !!}
+                {!! Form::submit('Create new category') !!}
```
Comment (nguyenthanhtung88) [inline resources/views/categories/home.blade.php]: Wrong Indent

File: resources/views/categories/home.blade.php [hunk 30]
```diff
@@ -0,0 +1,62 @@
...
+               {!! Form::close() !!}
```
Comment (nguyenthanhtung88) [inline resources/views/categories/home.blade.php]: Wrong Indent

File: resources/views/categories/home.blade.php [hunk 31]
```diff
@@ -0,0 +1,62 @@
+@extends('app')
+@section('title')
+    {{$title}}
+@endsection
+@section('content')
+    @if ($user->isAdmin())
+        <p>
+            {!! Form::open(['method' => 'get', 'route' => 'categories.create']) !!}
```
Comment (nguyenthanhtung88) [inline resources/views/categories/home.blade.php]: Why using `form` here +_+ it's not necessary.
You can use `link` is enough +_+

File: resources/views/categories/home.blade.php [hunk 32]
```diff
@@ -0,0 +1,62 @@
...
+                </tr>
+                <tr>
+                    <td>
+                        <p>{{ $category->description }}</p>
+                    </td>
+                </tr>
+            @if ($user->isAdmin())
+                    <tr>
+                        <td>
+                            {!! Form::open(['method' => 'get', 'route' => ['categories.edit', $category->id]]) !!}
```
Comment (nguyenthanhtung88) [inline resources/views/categories/home.blade.php]: Use `link` tag

File: resources/views/categories/edit.blade.php [hunk 33]
```diff
@@ -0,0 +1,49 @@
...
+                            ['style' => 'max-height: 100px;'])
+                        !!}
+                    </td>
+                @else
+                    <td rowspan='3'>
+                        {!! Html::image('images/categories/noimage.png', 'No image',
+                            ['style' => 'max-width: 100px;'])
+                        !!}
+                    </td>
+                @endif
```
Comment (nguyenthanhtung88) [inline resources/views/categories/edit.blade.php]: This `if` statement i suggest you to improve because it appear on `categories.index` view too. What i suggest is:

1. You should define new `config` for assets folder, in this case text `'images/categories/'` should be defined in config. Because maybe in the future, in your real project, folder for category images can be moved to another location => don't use fixed string for this folder in view, if changed, you have to change in multiple views +_+
2. You should try to define `Accessor` (https://laravel.com/docs/5.1/eloquent-mutators#accessors-and-mutators) in your `Category` model. In your views, only call `$category->image`, and in case of category has no image, Model will do it for you.

==> Please take time try to improve this code, your logic should be better :dancer:

File: app/Http/Controllers/CategoryController.php [hunk 34]
```diff
@@ -0,0 +1,87 @@
...
+        return redirect('/categories');
+    }
+
+    public function destroy(Request $request, $id)
+    {
+        try {
+            $category = Category::findOrFail($id);
+            $category->delete();
+            \Session::flash('flash_success', 'Delete successful!');
+        } catch     (ModelNotFoundException $e) {
```
Comment (nguyenthanhtung88) [inline app/Http/Controllers/CategoryController.php]: Too much space after `catch`

File: app/Http/Controllers/CategoryController.php [hunk 35]
```diff
@@ -0,0 +1,87 @@
...
+    public function edit($id)
+    {
+        try {
+            $category = Category::findOrFail($id);
+            return view('categories.edit', [
+                'category' => $category,
+                'title' => 'Edit category',
+                'user' => $this->user
+            ]);
+        } catch(ModelNotFoundException $e) {
```
Comment (nguyenthanhtung88) [inline app/Http/Controllers/CategoryController.php]: one space after `catch`

File: app/Category.php [hunk 36]
```diff
@@ -20,4 +22,30 @@ public function words()
...
     }
 
+    public function getImageAttribute($values)
+    {
+        if (!empty($values)) {
+            return $values;
+        }
+        else {
+            return 'noimage.png';
+        }
```
Comment (nguyenthanhtung88) [inline app/Category.php]: `return (!empty($values)) ? $values : 'noimage.png';` is enough.

File: app/Http/Controllers/CategoryController.php [hunk 37]
```diff
@@ -0,0 +1,87 @@
...
+        } catch (Exception $e) {
+            \Session::flash('flash_error', 'Category creation failed.');
+        }
+        return redirect('/categories');
+    }
+
+    public function update(Request $request, $id)
+    {
+        $this->validate($request, [
+                'category_name' => 'required|max:255',
```
Comment (nguyenthanhtung88) [inline app/Http/Controllers/CategoryController.php]: Back 1 indent

File: app/Http/Controllers/Controller.php [hunk 38]
```diff
@@ -10,4 +10,11 @@
 abstract class Controller extends BaseController
 {
     use AuthorizesRequests, DispatchesJobs, ValidatesRequests;
+
+    public $user;
```
Comment (nguyenthanhtung88) [inline app/Http/Controllers/Controller.php]: Should be `protected`

File: app/Http/Middleware/AdminMiddleware.php [hunk 39]
```diff
@@ -0,0 +1,25 @@
...
+     * @return mixed
+     */
+    public function handle($request, Closure $next)
+    {
+        if ($request->user()->isAdmin()) {
+            return $next($request);
+        }
+        else {
+            return redirect('/');
+        }
```
Comment (nguyenthanhtung88) [inline app/Http/Middleware/AdminMiddleware.php]: 1. I think we don't need `else` statement here, because if condition in `if` statement not valid, it will go outside.
```
if ($request->user()->isAdmin()) {
    return $next($request);
}

return redirect('/');
```
2. Why don't you implement this Middleware like `Authenticate` ?

File: resources/views/categories/edit.blade.php [hunk 40]
```diff
@@ -0,0 +1,39 @@
...
+            </tr>
+            <tr>
+                <td>
+                    {!! Form::file('category_image') !!}
+                </td>
+            </tr>
+        </table>
+        {!! Form::submit('Save') !!}
+    {!! Form::close() !!}
+    <a href="{{ url('categories/') }}" class="btn btn-default btn-lg" role="button">Back</a>
```
Comment (nguyenthanhtung88) [inline resources/views/categories/edit.blade.php]: Please use `HTML helper`

File: resources/views/categories/index.blade.php [hunk 41]
```diff
@@ -0,0 +1,52 @@
+@extends('app')
+@section('title')
+    {{ $title }}
+@endsection
+@section('content')
+    @if ($user->isAdmin())
+        <p>
+            {!! Form::open(['method' => 'get', 'route' => 'categories.create']) !!}
```
Comment (nguyenthanhtung88) [inline resources/views/categories/index.blade.php]: Please use `HTML Helper`, don't need open a new form GET just for a link, because when you click a link, it's same as create GET request +_+
https://laravelcollective.com/docs/5.0/html#generating-urls

File: resources/views/categories/edit.blade.php [hunk 42]
```diff
@@ -0,0 +1,39 @@
...
+    {{ $title }}
+@endsection
+@section('content')
+    {!! Form::open(['method' => 'patch', 'route' => ['categories.update', $category->id], 'files' => 'true']) !!}
+        <table>
+            <tr>
+                <td rowspan='3'>
+                    {!! Html::image(config()->get('paths.category_image') . $category->image, $category->name,
+                                ['style' => 'max-width: 100px;'])
+                            !!}
```
Comment (nguyenthanhtung88) [inline resources/views/categories/edit.blade.php]: Wrong Indent

File: resources/views/categories/edit.blade.php [hunk 43]
```diff
@@ -0,0 +1,39 @@
...
+@section('title')
+    {{ $title }}
+@endsection
+@section('content')
+    {!! Form::open(['method' => 'patch', 'route' => ['categories.update', $category->id], 'files' => 'true']) !!}
+        <table>
+            <tr>
+                <td rowspan='3'>
+                    {!! Html::image(config()->get('paths.category_image') . $category->image, $category->name,
+                                ['style' => 'max-width: 100px;'])
```
Comment (nguyenthanhtung88) [inline resources/views/categories/edit.blade.php]: Don't make inline style in view, create css class for this style and apply by class

File: resources/views/categories/index.blade.php [hunk 44]
```diff
@@ -0,0 +1,52 @@
...
+                </tr>
+                <tr>
+                    <td>
+                        <p>{{ $category->description }}</p>
+                    </td>
+                </tr>
+            @if ($user->isAdmin())
+                <tr>
+                    <td>
+                        {!! Form::open(['method' => 'get', 'route' => ['categories.edit', $category->id]]) !!}
```
Comment (nguyenthanhtung88) [inline resources/views/categories/index.blade.php]: Form with method `GET` is same as a link, please make a link here instead of opening a new form

File: resources/views/categories/index.blade.php [hunk 45]
```diff
@@ -0,0 +1,52 @@
...
+                        {!! Form::close() !!}
+                        {!! Form::open(['method' => 'delete', 'route' => ['categories.destroy', $category->id]]) !!}
+                            {!! Form::submit('Delete') !!}
+                        {!! Form::close() !!}
+                    </td>
+                </tr>
+            </table>
+            @else
+            </table>
+            @endif
```
Comment (nguyenthanhtung88) [inline resources/views/categories/index.blade.php]:
```
</table>
@else
</table>
@endif
```
=> Should change to:
```
@endif
</table>
```
==> You don't have to use `</table>` 2 times

File: app/Http/Controllers/UserController.php [hunk 46]
```diff
@@ -10,10 +10,30 @@
...
     {
         $title = trans('common.users.page_title');
-        $user = \Auth::user();
-        return view('home', compact('title', 'user'));
+        return view('home', ['user' => $this->user, 'title' => $title]);
+    }
+
+    public function profile(Request $request, $id)
+    {
+        $data['user'] = User::find($id);
```
Comment (nguyenthanhtung88) [inline app/Http/Controllers/UserController.php]: Use `findOrFail` in case of not found user

File: app/Http/Controllers/UserController.php [hunk 47]
```diff
@@ -10,10 +10,30 @@
...
+        return view('admin.profile', $data);
+    }
+
+    public function update(Request $request)
+    {
+        $this->validate($request, [
+            'name' => 'required|max:255',
+        ]);
+        $user_id = $request->input('user_id');
```
Comment (nguyenthanhtung88) [inline app/Http/Controllers/UserController.php]: Use `camelCase` for variable name

File: app/Http/Controllers/UserController.php [hunk 48]
```diff
@@ -10,10 +10,30 @@
...
+        $new_name = $request->input('name');
```
Comment (nguyenthanhtung88) [inline app/Http/Controllers/UserController.php]: `camelCase` for variable name

File: app/Http/Controllers/UserController.php [hunk 49]
```diff
@@ -10,10 +10,30 @@
...
+        $user = Users::find($user_id);
+        $user->name = $new_name;
+        $user_>save();
```
Comment (nguyenthanhtung88) [inline app/Http/Controllers/UserController.php]: why `_`  ?? 
Oops, this error i think you fixed before? why this PR include this ?

File: app/Http/routes.php [hunk 50]
```diff
@@ -1,26 +1,19 @@
...
     'auth' => 'Auth\AuthController',
     'password' => 'Auth\PasswordController',
 ]);
 
+Route::post('/update-name', 'UserController@update');
 
-Route::group(['middleware' => 'auth'], function () {
+Route::group(['middleware' => 'auth.admin'], function() {
+    Route::resource('categories', 'CategoryController');
+    Route::resource('words', 'WordController');
```
Comment (nguyenthanhtung88) [inline app/Http/routes.php]: You don't implement `WordController` in this PR, please remove it and add when you implement later

File: app/Http/routes.php [hunk 51]
```diff
@@ -1,26 +1,19 @@
...
-|
-*/
-
-
 Route::controllers([
     'auth' => 'Auth\AuthController',
     'password' => 'Auth\PasswordController',
 ]);
 
+Route::post('/update-name', 'UserController@update');
```
Comment (nguyenthanhtung88) [inline app/Http/routes.php]: 1. route name is not good
2. why we don't have any middleware for this :(

File: app/Http/Controllers/UserController.php [hunk 52]
```diff
@@ -10,10 +10,30 @@
...
+
     public function index()
     {
         $title = trans('common.users.page_title');
-        $user = \Auth::user();
-        return view('home', compact('title', 'user'));
+        return view('home', ['user' => $this->user, 'title' => $title]);
+    }
+
+    public function profile(Request $request, $id)
```
Comment (nguyenthanhtung88) [inline app/Http/Controllers/UserController.php]: Where's the route and where's the view? 
If you want to implement function for User (like view or update), please make another PR.
This PR is only for Category function.

File: resources/views/categories/edit.blade.php [hunk 53]
```diff
@@ -0,0 +1,39 @@
...
+            </tr>
+            <tr>
+                <td>
+                    {!! Form::file('category_image') !!}
+                </td>
+            </tr>
+        </table>
+        {!! Form::submit('Save') !!}
+    {!! Form::close() !!}
+    {!! link_to('categories', $title = 'Back') !!}
```
Comment (nguyenthanhtung88) [inline resources/views/categories/edit.blade.php]: Take a look at https://github.com/LaravelCollective/html/blob/5.2/src/helpers.php#L14
=> I think you don't need to put `$title = 'Back'` in second parameter, `link_to('categories', 'Back')` is enough

File: resources/views/categories/index.blade.php [hunk 54]
```diff
@@ -0,0 +1,44 @@
+@extends('app')
+@section('title')
+    {{ $title }}
+@endsection
+@section('content')
+    @if ($user->isAdmin())
+        <p>
+            {!! link_to_route('categories.create', $title = 'Create new category') !!}
```
Comment (nguyenthanhtung88) [inline resources/views/categories/index.blade.php]: `link_to_route('categories.create', 'Create new category')`

File: resources/views/categories/index.blade.php [hunk 55]
```diff
@@ -0,0 +1,44 @@
...
+                </tr>
+                <tr>
+                    <td>
+                        <p>{{ $category->description }}</p>
+                    </td>
+                </tr>
+                @if ($user->isAdmin())
+                <tr>
+                    <td>
+                        {!! link_to_route('categories.edit', $title = 'Edit', $parameters = [$category->id] ) !!}
```
Comment (nguyenthanhtung88) [inline resources/views/categories/index.blade.php]: Should be `link_to_route('categories.edit', 'Edit', [$category->id])` => and don't make one space before `)`

File: resources/views/categories/index.blade.php [hunk 56]
```diff
@@ -0,0 +1,44 @@
...
+                        {!! link_to_route('categories.destroy', $title = 'Delete', $parameters = [$category->id]) !!}
```
Comment (nguyenthanhtung88) [inline resources/views/categories/index.blade.php]: `link_to_route('categories.destroy', 'Delete', [$category->id])`

File: resources/views/categories/index.blade.php [hunk 57]
```diff
@@ -0,0 +1,46 @@
...
+                </tr>
+                <tr>
+                    <td>
+                        <p>{{ $category->description }}</p>
+                    </td>
+                </tr>
+            @if ($user->isAdmin())
+                <tr>
+                    <td>
+                        {!! link_to_route('categories.edit', 'Edit', $parameters = [$category->id]) !!}
```
Comment (nguyenthanhtung88) [inline resources/views/categories/index.blade.php]: I think we don't need `$parameters =`

**Comment by nguyenthanhtung88** [comment]
OK
